OpenRA.Mods.Common/Traits/EjectOnDeath.cs

```diff
@@ -50,58 +50,55 @@ void INotifyKilled.Killed(Actor self, AttackInfo e)
 			if (IsTraitDisabled || self.Owner.WinState == WinState.Lost || !self.World.Map.Contains(self.Location))
 				return;
 
-			var r = self.World.SharedRandom.Next(1, 100);
-
-			if (r <= 100 - Info.SuccessRate)
+			if (self.World.SharedRandom.Next(100) >= Info.SuccessRate)
 				return;
 
 			var cp = self.CenterPosition;
 			var inAir = !self.IsAtGroundLevel();
 			if ((inAir && !Info.EjectInAir) || (!inAir && !Info.EjectOnGround))
 				return;
 
-			var pilot = self.World.CreateActor(false, Info.PilotActor.ToLowerInvariant(),
-				new TypeDictionary { new OwnerInit(self.Owner), new LocationInit(self.Location) });
-
-			var pilotPositionable = pilot.TraitOrDefault<IPositionable>();
-			var pilotCell = self.Location;
-			var pilotSubCell = pilotPositionable.GetAvailableSubCell(pilotCell);
-			if (pilotSubCell == SubCell.Invalid)
+			if (!Info.AllowUnsuitableCell)
 			{
-				if (!Info.AllowUnsuitableCell)
-				{
-					pilot.Dispose();
+				var pilotInfo = self.World.Map.Rules.Actors[Info.PilotActor.ToLowerInvariant()];
+				var pilotPositionable = pilotInfo.TraitInfo<IPositionableInfo>();
+				if (!pilotPositionable.CanEnterCell(self.World, null, self.Location))
 					return;
-				}
-
-				pilotSubCell = SubCell.Any;
 			}
 
+			var td = new TypeDictionary
+			{
+				new OwnerInit(self.Owner),
+				new LocationInit(self.Location),
+			};
+
+			// If airborne, offset the spawn location so the pilot doesn't drop on another infantry's head
+			var spawnPos = cp;
 			if (inAir)
 			{
-				self.World.AddFrameEndTask(w =>
+				var subCell = self.World.ActorMap.FreeSubCell(self.Location);
+				if (subCell != SubCell.Invalid)
```

Comment:
Shouldn't we skip actor creation if `!Info.AllowUnsuitableCell` and `subCell == SubCell.Invalid`?

Reply:
We do, several lines above this.

Reply:
I see, this is just to make sure we don't use `SubCell.Invalid` if we allow unsuitable cells.